### DISCUSSION TITLE
Safe conversion of numpy array to ndarray without copying data

### DISCUSSION THREAD

**Discussion by peendebak** [rust-ndarray/ndarray#1179] [Q&A]
From Python I am trying to convert an object into the following structure:
``` 
pub enum ArrayLike {
    Float(f64),
    Int(isize),
    FloatArray(ArrayD<f64>), // todo: replace with ArrayViewD<a', f64>
}
```
I have this working, with the following definition and the standard `pyo3` setup:
```
impl<'source> FromPyObject<'source> for ArrayLike {
    fn extract(ob: &'source PyAny) -> PyResult<Self> {
        if let Ok(f) = <PyFloat as pyo3::PyTryFrom>::try_from(ob) {
            return Ok(ArrayLike::Float(f.extract()?));
        } else if let Ok(i) = <PyLong as pyo3::PyTryFrom>::try_from(ob) {
            return Ok(ArrayLike::Float(i.extract()?));
        } else if let Ok(a) = <PyArrayDyn<f64> as pyo3::PyTryFrom>::try_from(ob) {
            let b = unsafe { a.as_array() };
            return Ok(ArrayLike::FloatArray(b.to_owned()));
        } else {
             ....
```
Note that there is an `unsafe` block for the numpy array conversion. The documentation of [`as_array`](https://docs.rs/numpy/latest/numpy/array/struct.PyArray.html#method.to_array) suggests to either use a `PyReadonlyArray` or the `to_array` method.

1) I cannot find the `to_array` method on the documentation for latest. Does it exist, or is the documentation old?
2) I tried to use the `PyReadonlyArray` by changing the code to
```
        } else if let Ok(a) = <PyReadonlyArrayDyn<f64> as pyo3::PyTryFrom>::try_from(ob) {
```
but this results in `the trait 'PyNativeType'  is not implemented for 'PyReadonlyArray<'_, f64, Dim<IxDynImpl>>'`. Any suggestions on how to use `PyReadonlyArray`?

**Comment by adamreichold** [comment]
I think this discussion is probably more relevant to the `rust-numpy` crate providing the bindings between NumPy and `ndarray` than to `ndarray` itself.

In any case, maybe the example provided [here](https://github.com/PyO3/rust-numpy/blob/main/examples/simple/src/lib.rs#L94) is helpful for you as an example of how to implement polymorphism like use the derive macro for `FromPyObject`.

Concerning the question on `PyReadonlyArray`, it is a wrapper around `PyArray` and hence not itself a native Python type. It does [implement `FromPyObject`](https://docs.rs/numpy/latest/numpy/struct.PyReadonlyArray.html#impl-FromPyObject%3C%27py%3E), i.e. you should be able to use instead of `FromPyObject::extract` instead of `PyTryFrom::try_from`. Or something like

```rust
#[derive(FromPyObject)]
pub enum ArrayLike<'a> {
    Float(f64),
    Int(isize),
    FloatArray(PyReadonlyArrayD<'a, f64>),
}
```
